### PR TITLE
[Android] welcome onboarding

### PR DESCRIPTION
**Pull request by tapanmodh**
Resolves https://github.com/brave/brave-browser/issues/22946

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one https://github.com/brave/security/issues/871
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 1]
```diff
@@ -98,13 +106,25 @@ public BlockersInfo() {
...
 
         public int mAdsBlocked;
         public int mTrackersBlocked;
         public int mHTTPSUpgrades;
         public int mScriptsBlocked;
         public int mFingerprintsBlocked;
+        public ArrayList<String> mBlockerNames;
+    }
+
+    private static class DisconnectEntity {
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:120]: What is the meaning of disconnect entity? I didn't find the description in the docs and in the issue.

Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:120]: It is to identify the ads/tracker to the big tech company. we are using the same json file in the ios as well.

Comment (AlexeyBarabash) [reply android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:120]: thanks

File: android/java/org/chromium/chrome/browser/firstrun/BraveFreIntentCreator.java [hunk 2]
```diff
@@ -5,18 +5,12 @@
...
-import android.app.Activity;
-import android.app.PendingIntent;
 import android.content.Context;
 import android.content.Intent;
 
 import androidx.annotation.Nullable;
 
-import org.chromium.base.IntentUtils;
-import org.chromium.chrome.browser.BraveConfig;
-import org.chromium.chrome.browser.browserservices.intents.BrowserServicesIntentDataProvider;
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/firstrun/BraveFreIntentCreator.java:17]: :+1: for cleaning unused imports

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 3]
```diff
@@ -170,15 +191,87 @@ public BraveShieldsHandler(Context context) {
...
+        }
+    }
+
+    private String loadDisconnectEntityJSONFromAsset() {
+        String json = null;
+        try {
+            InputStream is = mContext.getAssets().open("disconnect_entitylist.json");
+            int size = is.available();
+            byte[] buffer = new byte[size];
+            is.read(buffer);
```
[the remark this answers is not in the document]
Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java]: done

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 4]
```diff
@@ -114,7 +134,8 @@ public BlockersInfo() {
     private View mHardwareButtonMenuAnchor;
     private final Map<Integer, BlockersInfo> mTabsStat =
         Collections.synchronizedMap(new HashMap<Integer, BlockersInfo>());
-
+    private Map<String, DisconnectEntity> mDisconnectEntity =
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:137]: I see the map is used, but the map is then iterated element by element in `getBlockerCompanyName`. Nowhere is used for fast search of DisconnectEntity. So there are no advantages of using the map here. 

I would suggest to replace it with some linear container as `ArrayList`. 

WDYT?

File: android/java/org/chromium/chrome/browser/util/BraveConstants.java [hunk 5]
```diff
@@ -16,4 +16,8 @@ public final class BraveConstants {
 
     // used in BravseSetDefaultBrowserUtils
     public static final int DEFAULT_BROWSER_ROLE_REQUEST_CODE = 37;
+
+    public static final String BRAVE_PRODUCTION_PACKAGE_NAME = "com.brave.browser";
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/util/BraveConstants.java:20]: ++ for moving it here

File: android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+import java.lang.Math;
+
+public class WelcomeOnboardingActivity extends FirstRunActivityBase {
+    // mInitializeViewsDone and mInvokePostWorkAtInitializeViews are accessed
+    // from the same thread, so no need to use extra locks
+    private static final String P3A_URL = "https://brave.com/privacy/browser/#how-we-improve-brave";
+    private boolean mInitializeViewsDone;
+    private boolean mInvokePostWorkAtInitializeViews;
+    private boolean mIsP3aEnabled;
+    private boolean isTablet;
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java:50]: nit: naming for private member

File: android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        mBtnNegative.setOnClickListener(view -> {
+            if (mCurrentStep == 2) {
+                CustomTabActivity.showInfoPage(this, P3A_URL);
+            } else {
+                nextOnboardingStep();
+            }
+        });
+    }
+
+    private void startTimer(int timer) {
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java:146]: maybe rename `timer` to `delayMillis`?

File: android/java/org/chromium/chrome/browser/onboarding/v2/HighlightView.java [hunk 8]
```diff
@@ -7,54 +7,147 @@
...
 
+    public void initializeAnimators() {
+        mAnimatorSet = new AnimatorSet();
+
+        ObjectAnimator scaleXAnimator = ObjectAnimator.ofFloat(
+                this, "innerRadius", mInnerRadius * 0.7f, mInnerRadius * 1.1f);
+        scaleXAnimator.setRepeatCount(ValueAnimator.INFINITE);
+
+        ObjectAnimator scaleBigAnimator = ObjectAnimator.ofFloat(this, "outerRadius",
+                mOuterRadius * 0.9f, mOuterRadius * 1.2f); // 1.2f, mOuterRadius * 1.4f);
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/onboarding/v2/HighlightView.java:112]: Is this comment a leftover?

**Review by AlexeyBarabash — CHANGES_REQUESTED** [review]
I have put some comments and questions.

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 9]
```diff
@@ -1024,9 +1063,6 @@ private void checkForNotificationData() {
                     checkForBraveStats();
                     break;
                 case RetentionNotificationUtil.DAY_6:
-                case RetentionNotificationUtil.BRAVE_STATS_ADS_TRACKERS:
-                case RetentionNotificationUtil.BRAVE_STATS_DATA:
-                case RetentionNotificationUtil.BRAVE_STATS_TIME:
```
Comment (alexsafe) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1029]: good to see these out

File: android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+        mTvWelcome = findViewById(R.id.tv_welcome);
+        mTvCard = findViewById(R.id.tv_card);
+        mTvDefault = findViewById(R.id.tv_default);
+        mCheckboxCrash = findViewById(R.id.checkbox_crash);
+        mCheckboxP3a = findViewById(R.id.checkbox_p3a);
+        mBtnPositive = findViewById(R.id.btn_positive);
+        mBtnNegative = findViewById(R.id.btn_negative);
+
+        int margin;
+        if (isTablet) {
```
Comment (alexsafe) [inline android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java:107]: @tapanmodh nit could be replaced with 
`int margin = isTablet ? 200 : 50;`

File: android/java/org/chromium/chrome/browser/onboarding/v2/HighlightView.java [hunk 11]
```diff
@@ -7,54 +7,147 @@
...
 public class HighlightView extends FrameLayout {
 
     private static final int ALPHA_60_PERCENT = 153;
+    private static final int DEFAULT_ANIMATION_DURATION = 1000;
 
     private Paint eraserPaint;
     private Paint basicPaint;
 
     private HighlightItem item;
+    private int itemWidth;
```
Comment (alexsafe) [inline android/java/org/chromium/chrome/browser/onboarding/v2/HighlightView.java:42]: @tapanmodh private members naming

File: android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java (hunk 10 above)
Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java:107]: right, will do it.

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 12]
```diff
@@ -98,23 +106,35 @@ public BlockersInfo() {
...
 
         public int mAdsBlocked;
         public int mTrackersBlocked;
         public int mHTTPSUpgrades;
         public int mScriptsBlocked;
         public int mFingerprintsBlocked;
+        public ArrayList<String> mBlockerNames;
     }
 
+    /*private static class DisconnectEntity {
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:120]: Why do we need these commented lines?

Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:120]: yes, will remove it

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 13]
```diff
@@ -114,7 +122,7 @@ public BlockersInfo() {
     private View mHardwareButtonMenuAnchor;
     private final Map<Integer, BlockersInfo> mTabsStat =
         Collections.synchronizedMap(new HashMap<Integer, BlockersInfo>());
-
+    private Map<String, String> mDisconnectEntity = new HashMap<String, String>();
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:125]: Maybe rename `mDisconnectEntity` to `mDisconnectEntities` or `mResourceToCompanyName`
to more accurate reflect the meaning?

**Review by AlexeyBarabash — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+            mTvCard.setText(getResources().getString(R.string.p3a_title));
+            mBtnPositive.setText(getResources().getString(R.string.continue_text));
+            mBtnNegative.setText(getResources().getString(R.string.learn_more_onboarding));
+
+            if (PackageUtils.isFirstInstall(this)
+                    && (getPackageName().equals(BraveConstants.BRAVE_NIGHTLY_PACKAGE_NAME)
+                            || getPackageName().equals(BraveConstants.BRAVE_BETA_PACKAGE_NAME))
+                    && !OnboardingPrefManager.getInstance().isP3aCrashReportingMessageShown()) {
+                mCheckboxCrash.setChecked(true);
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java:201]: this is not correct. The value should always default to `isUsageAndCrashReportingPermittedByUser` as specified https://github.com/brave/reviews/issues/1132#issuecomment-1335818255 and in our privacy policy

Comment (bridiver) [reply android/java/org/chromium/chrome/browser/firstrun/WelcomeOnboardingActivity.java:201]: actually sorry, this predates that, but this code has not changed so it is not incorrect

File: android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java [hunk 15]
```diff
@@ -170,15 +180,59 @@ public BraveShieldsHandler(Context context) {
...
+                }
+            }
+            isDisconnectEntityLoaded = true;
+
+        } catch (JSONException exception) {
+            exception.printStackTrace();
+        }
+    }
+
+    private String loadDisconnectEntityJSONFromAsset() {
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/shields/BraveShieldsHandler.java:207]: this is not right, we should not be doing this in java and we should not be embedding an adblock list directly in the code anywhere.

**Pull request merged**